### PR TITLE
[Improvements] Improve allgather sharded_state_dict compatibility and enable gate overlap

### PR DESCRIPTION
**Pull request by adam-xiaoyao**
### PR Category

Distributed Strategy

### PR Types

Improvements

### Description

Improve allgather sharded_state_dict compatibility and enable gate overlap

### 是否引起精度变化

否

### CODE REVIEW

File: src/paddlefleet/transformer/moe/moe_expert.py [hunk 1]
```diff
@@ -390,36 +415,62 @@ def _get_intermediate_sharded_state_dict(
...
+            w2_axis = 1  # [E, I_local, H] — shard intermediate dim
+        else:
+            # 2-D reshape: [E, H, 2*I_local] -> [E*H, 2*I_local]
+            w1 = w1_raw.reshape([E * H, 2 * I_local])
+            w1.name = self.weight1.name
+            # 2-D reshape: [E, I_local, H] -> [E*I_local, H]
+            w2 = w2_raw.reshape([E * I_local, H])
+            w2.name = self.weight2.name
+            w1_axis = 1  # [E*H, 2*I_local] — shard intermediate dim
+            w2_axis = 0  # [E*I_local, H] — shard expert×intermediate dim
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/moe/moe_expert.py]: P1 **优先级：P1**

**处理要求：请针对该评论修复并提交新的 commit。**

**问题：** 这里把 `weight1` 的 `[gate_local, up_local]` 直接展平成 2D 后用 `axis=1` 分片，同时把 `weight2` 展平成 `[E * I_local, H]` 后用 `axis=0` 分片；但 `shard_weight` 只能把每个 rank 的本地张量描述成一个连续的 global slice。以 `weight1` 为例，EP=2 时当前 metadata 组出的 global 列顺序是 `[gate_rank0, up_rank0, gate_rank1, up_rank1]`，而标准 2D 权重的逻辑顺序是 `[gate_rank0, gate_rank1, up_rank0, up_rank1]`。`weight2` 也会变成 rank-major（rank0 的所有 expert I 片段后接 rank1）而不是标准的 expert-major。

**影响：** checkpoint 的 `global_shape` 看起来能对齐，但 DCP reshard/load 会把 gate/up 或 expert/intermediate 片段放到错误位置；allgather checkpoint 跨 EP 度加载，或与 deepep 2D checkpoint 互转时会得到 shape 正确但数值布局错误的权重。当前注释也说明 AOA 还未实现，因此 metadata 里没有额外重排来修正这个布局。

**期望：** 在实现 AOA/custom sharded metadata 之前，不要用这个 2D flatten + single-axis shard 作为兼容格式；请保留显式 gate/up 和 intermediate 维度，或实现对应重排，并增加带可区分 sentinel 值的保存/加载顺序回归测试。当前安全形态至少应保持类似下面的布局：

```python
w1 = w1_raw.reshape([E, H, 2, I_local])
w1.name = self.weight1.name
w2 = w2_raw
w2.name = self.weight2.name
w1_axis = 3
w2_axis = 1
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
发现一个需要先修复的 checkpoint 布局问题，具体说明已放在 inline review comment 中。其余变更我没有看到需要阻塞的代码问题；本地环境缺少 `paddle`，只做了 Python 语法检查，GitHub 上仍有部分测试在运行中。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/moe/moe_expert.py (hunk 1 above)
Comment (adam-xiaoyao) [reply src/paddlefleet/transformer/moe/moe_expert.py]: OK, 我修改。

Comment (risemeup1111) [reply src/paddlefleet/transformer/moe/moe_expert.py]: 收到。当前 head 仍是上一轮已评审的提交，还没有包含修复；这条 P1 需要通过新的 commit 修复后我再复查确认。

File: tests/single_card_tests/ai_edited_test/moe/test_ai_allgather_dispatcher.py [hunk 2]
```diff
@@ -1463,107 +1507,146 @@ def test_int64_dtype_cast(self):
...
+            t.grouped_gemm_param = False
+            calls.append((key, axis))
+            return t
 
-        expert = self._FakeExpert(
-            intermediate_size_per_partition=8, ep_group=None
-        )
-        with self.assertRaises(ValueError):
-            GroupedMLPExpert.sharded_state_dict(expert)
+        with patch.object(me, "shard_weight", side_effect=_fake):
```
[the remark this answers is not in the document]
Comment (adam-xiaoyao) [reply tests/single_card_tests/ai_edited_test/moe/test_ai_allgather_dispatcher.py:1569]: OK，我修改。

File: tests/single_card_tests/ai_edited_test/moe/test_ai_allgather_dispatcher.py [hunk 3]
```diff
@@ -1463,107 +1507,219 @@ def test_int64_dtype_cast(self):
...
+        I_local = I_full // ep_size
+
+        for model_type in ("none", "qwen3_vl", "qwen3_5"):
+            e = self._make_expert(
+                ep_group=_mock_group(ep_size),
+                model_type=model_type,
+                i_local=I_local,
+                moe_int=I_full,
+            )
+            res = e._get_intermediate_sharded_state_dict(e.state_dict(), "p.")
```
Comment (risemeup1111) [inline tests/single_card_tests/ai_edited_test/moe/test_ai_allgather_dispatcher.py:1677]: P1 **优先级：P1**

**处理要求：请针对该评论修复并提交新的 commit。**

**问题：** 这个新增用例开始调用真实 `shard_weight`，但 `_mock_group(2)` 只设置了 `nranks/world_size/id`，没有设置 `rank`。DCP 的 `shard_weight` 会读取 `group.rank` 计算 `global_offset`，因此这里 `sw1.global_offset[3]` 实际会是 `MagicMock * I_local` 的结果，而不是 `0`，新增测试会失败。

**影响：** 修复 checkpoint 布局的回归用例无法稳定通过，CI 会被测试失败阻塞。

**期望：** 给 fake group 补齐真实 `shard_weight` 需要的 `rank` 字段，并最好允许测试覆盖非 0 rank 的 offset。最小修复可以放在 helper 里：

```python
def _mock_group(nranks=1, rank=0):
    """Create a mock distributed group."""
    g = MagicMock()
    g.nranks = nranks
    g.world_size = nranks
    g.rank = rank
    g.id = 0
    return g
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查新的提交：之前指出的 allgather checkpoint 2D flatten 布局问题已经改回 4D metadata，原 P1 认为已解决。

不过新增的真实 `shard_weight` metadata 测试里 fake group 缺少 `rank`，会导致新增断言失败；具体见新的 inline comment。CI 目前也仍在跑。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by zjjlivein — APPROVED** [review]

**Review by risemeup1111 — APPROVED** [review]
已复查当前提交：之前两个 P1 问题都已经通过后续代码变更解决。

- allgather checkpoint metadata 已保持 4D/3D 的中间维分片布局。
- `_mock_group` 已补齐 `rank`，并增加了非 0 rank 的 `global_offset` 回归覆盖；该线程里的测试问题已解决。

本地可做的 Python 语法检查已通过；当前仍有 CI/仓库审批状态需要以 GitHub checks 为准。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by ForFishes — APPROVED** [review]
LGTM
